File: frontend/ksweb-app/src/app/pages/ordens-servico/ordens-servico.models.ts
export interface OrdemServicoListItem {
  numeroOs: number;
  titulo: string;
  criadoEmMs: number;
  criadoEm: string;
  solicitanteId: number;
  solicitante: string;
  criadoPorId: number;
  criadoPor: string;
  responsavelId: number;
  responsavel: string;
  setorId: number;
  setor: string;
  statusId: number;
  statusNome: string;
  atrasada: boolean;
  lida: boolean;
}

export interface OrdemServicoListResponse {
  items: OrdemServicoListItem[];
  total: number;
  page: number;
  pageSize: number;
  usuarioLogadoNome: string | null;
}

export interface OrdemServicoStatusOption {
  statusId: number;
  statusName: string;
}

export interface OrdemServicoUsuarioOption {
  userId: number;
  firstName: string;
}

export interface OrdemServicoFiltrosResponse {
  status: OrdemServicoStatusOption[];
  usuarios: OrdemServicoUsuarioOption[];
}

export interface OrdemServicoFiltro {
  page: number;
  pageSize: number;
  numero?: string;
  texto?: string;
  statusId?: string;
  filtrarPor?: 'solicitante' | 'criadoPor' | 'responsavel';
  filtroUsuarioNome?: string;
  filtroPessoa?: string;
  listarTudo?: boolean;
  usuarioId?: string;
}

export interface OrdemServicoGrupoOption {
  grpId: number;
  grpNome: string;
}

export interface OrdemServicoUsuarioGrupoOption {
  usrCodigo: number;
  usrNome: string;
}

export interface OrdemServicoCombosResponse {
  clientes: OrdemServicoUsuarioOption[];
  status: OrdemServicoStatusOption[];
  grupos: OrdemServicoGrupoOption[];
}

export interface OrdemServicoHistoricoItem {
  historyId: number;
  workorderId: number;
  operationOwnerId: number | null;
  operationTime: number | null;
  description: string | null;
  operation: string | null;
}

export interface OrdemServicoResolucaoItem {
  historyDiffId: number;
  historyId: number;
  columnName: string;
  prevValue: string | null;
  currentValue: string | null;
  operationTime: number | null;
}

export interface OrdemServicoFormResponse {
  workorderId: number | null;
  requesterId: number | null;
  requesterName: string | null;
  createdById: number | null;
  createdByName: string | null;
  createdTime: number | null;
  title: string;
  description: string;
  fullDescription: string;
  ownerId: number | null;
  ownerName: string | null;
  statusId: number | null;
  statusName: string | null;
  queueId: number | null;
  queueName: string | null;
  lastResolution: string | null;
  historico: OrdemServicoHistoricoItem[];
  resolucoes: OrdemServicoResolucaoItem[];
}

export interface SalvarOrdemServicoRequest {
  requesterId: number;
  title: string;
  description: string;
  fullDescription: string;
  ownerId: number | null;
  statusId: number | null;
}

export interface SalvarResolucaoRequest {
  html: string;
  texto: string;
}

export interface AlterarStatusAtribuirRequest {
  statusId: number | null;
  queueId: number | null;
  ownerId: number | null;
}

File: backend/KSWeb.Api/Models/OrdensServicoModels.cs
namespace KSWeb.Api.Models;

public sealed record OrdemServicoListItem(
    long OsCodigo,
    long OsData,
    string OsTitulo,
    string Cliente,
    string UsuarioAbertura,
    string? Status,
    bool? IsRead,
    string UsuarioResponsavel);

public sealed record OrdemServicoListResponse(
    IReadOnlyList<OrdemServicoListItem> Items,
    int Total,
    int Page,
    int PageSize,
    string? UsuarioLogadoNome);

public sealed record OrdemServicoStatusOption(
    long StatusId,
    string StatusName);

public sealed record OrdemServicoUsuarioOption(
    long UserId,
    string FirstName);

public sealed record OrdemServicoFiltrosResponse(
    IReadOnlyList<OrdemServicoStatusOption> Status,
    IReadOnlyList<OrdemServicoUsuarioOption> Usuarios);

public sealed record OrdemServicoGrupoOption(
    long GrpId,
    string GrpNome);

public sealed record OrdemServicoUsuarioGrupoOption(
    long UsrCodigo,
    string UsrNome);

public sealed record OrdemServicoCombosResponse(
    IReadOnlyList<OrdemServicoUsuarioOption> Clientes,
    IReadOnlyList<OrdemServicoStatusOption> Status,
    IReadOnlyList<OrdemServicoGrupoOption> Grupos);

public sealed record OrdemServicoFormResponse(
    long? WorkorderId,
    long? RequesterId,
    string? RequesterName,
    long? CreatedById,
    string? CreatedByName,
    long? CreatedTime,
    string Title,
    string Description,
    string FullDescription,
    long? OwnerId,
    string? OwnerName,
    long? StatusId,
    string? StatusName,
    long? QueueId,
    string? QueueName,
    string? LastResolution,
    IReadOnlyList<OrdemServicoHistoricoItem> Historico,
    IReadOnlyList<OrdemServicoResolucaoItem> Resolucoes);

public sealed record OrdemServicoHistoricoItem(
    long HistoryId,
    long WorkorderId,
    long? OperationOwnerId,
    long? OperationTime,
    string? Description,
    string? Operation);

public sealed record OrdemServicoResolucaoItem(
    long HistoryDiffId,
    long HistoryId,
    string ColumnName,
    string? PrevValue,
    string? CurrentValue,
    long? OperationTime);

public sealed record SalvarOrdemServicoRequest(
    long RequesterId,
    string Title,
    string Description,
    string FullDescription,
    long? OwnerId,
    long? StatusId);

public sealed record SalvarResolucaoRequest(string Html, string Texto);

public sealed record AlterarStatusAtribuirRequest(
    long? StatusId,
    long? QueueId,
    long? OwnerId);
